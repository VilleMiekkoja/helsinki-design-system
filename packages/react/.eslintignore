**/*.d.ts
webpack.config.js
.templates
lib/
