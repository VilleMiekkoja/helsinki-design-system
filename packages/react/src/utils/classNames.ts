import classnames from 'classnames/dedupe';

export default classnames;
