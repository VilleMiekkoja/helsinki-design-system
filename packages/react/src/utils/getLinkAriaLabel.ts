import React from 'react';

export default (
  ariaLabel: string,
  label: React.ReactNode,
  openInNewTabAriaLabel?: string,
  openInExternalDomainAriaLabel?: string,
  external?: boolean,
  openInNewTab?: boolean,
) => {
  let childrenText = ariaLabel || label;
  const newTabText = openInNewTab ? openInNewTabAriaLabel || 'Avautuu uudessa välilehdessä.' : '';
  const externalText = external ? openInExternalDomainAriaLabel || 'Siirtyy toiseen sivustoon.' : '';

  if (childrenText && childrenText.slice(-1) !== '.') {
    childrenText = `${childrenText}.`;
  }

  return [childrenText, newTabText, externalText].filter((text) => text).join(' ');
};
