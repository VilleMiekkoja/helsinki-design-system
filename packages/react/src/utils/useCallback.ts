import { EventHandler, useCallback } from 'react';

export const useCallbackIfDefined = (callback) => {
  const handler = (event) => {
    if (callback) callback(event);
  };

  return useCallback(handler, [callback]);
};

export const useEnterOrSpacePressCallback = (callback) => {
  const handler = (event) => {
    if (event.key === 'Enter' || event.key === ' ') callback(event);
  };

  return useCallbackIfDefined(handler);
};

export const withDefaultPrevented = (callback: EventHandler<never>) =>
  function patchedCallback(event) {
    event.preventDefault();
    return callback.call(callback, event);
  };

export const withPropagationStopping = (callback: EventHandler<never>) =>
  function patchedCallback(event) {
    event.stopPropagation();
    return callback.call(callback, event);
  };
