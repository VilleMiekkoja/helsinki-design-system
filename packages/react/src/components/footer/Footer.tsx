import React from 'react';

// import core base styles
import 'hds-core';
import styles from './Footer.module.scss';
import { Koros, KorosType } from '../koros';
import classNames from '../../utils/classNames';
import { FooterNavigation } from './footerNavigation/FooterNavigation';
import { FooterNavigationGroup } from './footerNavigationGroup/FooterNavigationGroup';
import { FooterNavigationHeading } from './footerNavigationHeading/FooterNavigationHeading';
import { FooterNavigationLink } from './footerNavigationLink/FooterNavigationLink';
import { FooterUtilities } from './footerUtilities/FooterUtilities';
import { FooterUtilityGroup } from './footerUtilityGroup/FooterUtilityGroup';
import { FooterSoMe } from './footerSoMe/FooterSoMe';
import { FooterBase } from './footerBase/FooterBase';
import { FooterCustom } from './footerCustom/FooterCustom';
import { FooterTheme } from './Footer.interface';
import { getComponentFromChildren } from '../../utils/getChildren';
import { useTheme } from '../../hooks/useTheme';

export type FooterProps = React.PropsWithChildren<{
  /**
   * Additional class names to apply to the footer
   */
  className?: string;
  /**
   * Props that will be passed to the native `<footer>` element
   */
  footerProps?: React.ComponentPropsWithoutRef<'footer'>;
  /**
   * Koros type to use in the footer
   */
  korosType?: KorosType;
  /**
   * Defines the footer theme
   */
  theme?: FooterTheme;
  /**
   * The title of the service shown on top of the the footer
   */
  title?: React.ReactNode;
}>;

export const Footer = ({
  children,
  className,
  footerProps,
  korosType = 'basic',
  theme = 'light',
  title,
}: FooterProps) => {
  // custom theme class that is applied to the root element
  const customThemeClass = useTheme<FooterTheme>(styles.footer, theme);
  // filter out navigation from other children so that they can be rendered separately
  const [navigation, childrenWithoutNavigation] = getComponentFromChildren(children, 'FooterNavigation');

  return (
    <footer
      {...footerProps}
      className={classNames(
        styles.footer,
        styles[`koros-${korosType}`],
        typeof theme === 'string' && styles[`theme-${theme}`],
        customThemeClass,
        className,
      )}
    >
      <Koros className={classNames(styles.koros, styles[korosType])} type={korosType} />
      <div className={styles.footerContent}>
        <section className={classNames(styles.navigationContainer, styles.centeredNavigation)}>
          <div className={styles.titleWrapper}>{title && <h2 className={styles.title}>{title}</h2>}</div>
          {navigation}
        </section>
        {childrenWithoutNavigation}
      </div>
    </footer>
  );
};

Footer.Navigation = FooterNavigation;
Footer.NavigationGroup = FooterNavigationGroup;
Footer.NavigationHeading = FooterNavigationHeading;
Footer.NavigationLink = FooterNavigationLink;
Footer.Utilities = FooterUtilities;
Footer.UtilityGroup = FooterUtilityGroup;
Footer.SoMe = FooterSoMe;
Footer.Base = FooterBase;
Footer.Custom = FooterCustom;
