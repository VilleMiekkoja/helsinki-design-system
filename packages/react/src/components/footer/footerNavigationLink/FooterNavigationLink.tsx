import React from 'react';

// import core base styles
import 'hds-core';
import styles from './FooterNavigationLink.module.scss';
import { MergeElementProps } from '../../../common/types';
import classNames from '../../../utils/classNames';
import getLinkAriaLabel from '../../../utils/getLinkAriaLabel';
import { IconAngleRight, IconLinkExternal } from '../../../icons';
import { FooterVariant } from '../Footer.interface';

type ItemProps<Element> = React.PropsWithChildren<{
  /**
   * aria-label for providing detailed information for screen readers about a link.
   */
  ariaLabel?: string;
  /**
   * Element type
   */
  as?: Element;
  /**
   * Boolean indicating whether the link will lead user to external domain.
   */
  external?: boolean;
  /**
   * Icon placed on the left side of the item label
   */
  icon?: React.ReactNode;
  /**
   * Boolean indicating whether the link will open in new tab or not.
   */
  openInNewTab?: boolean;
  /**
   * The aria-label for opening link in a new tab
   */
  openInNewTabAriaLabel?: string;
  /**
   * The aria-label for opening link in an external domain
   */
  openInExternalDomainAriaLabel?: string;
  /**
   * The label for the item.
   */
  label?: string;
  /**
   * Whether the item should be displayed as an sub item in a sitemap item group
   * @internal
   */
  subItem?: boolean;
  /**
   * Internal variant to change styles based on context.
   * @internal
   */
  variant?: FooterVariant.Navigation | FooterVariant.Utility;
}>;

export type FooterNavigationLinkProps<Element extends React.ElementType = 'a'> = MergeElementProps<
  Element,
  ItemProps<Element>
>;

export const FooterNavigationLink = <T extends React.ElementType = 'a'>({
  ariaLabel,
  as,
  className,
  icon,
  external = false,
  openInNewTab = false,
  openInExternalDomainAriaLabel,
  openInNewTabAriaLabel,
  label,
  subItem = false,
  variant,
  ...rest
}: FooterNavigationLinkProps<T>) => {
  const Item: React.ElementType = as;
  const composedAriaLabel = getLinkAriaLabel(
    ariaLabel,
    label,
    openInNewTabAriaLabel,
    openInExternalDomainAriaLabel,
    external,
    openInNewTab,
  );

  return (
    // @ts-ignore
    <Item
      aria-label={ariaLabel}
      className={classNames(styles.item, subItem && styles.subItem, variant && styles[variant], className)}
      external={external}
      openInNewTab={openInNewTab}
      openInExternalDomainAriaLabel={openInExternalDomainAriaLabel}
      openInNewTabAriaLabel={openInNewTabAriaLabel}
      {...(openInNewTab && { target: '_blank', rel: 'noopener' })}
      {...((openInNewTab || external || ariaLabel) && { 'aria-label': composedAriaLabel })}
      {...rest}
    >
      {icon}
      {subItem && <IconAngleRight className={styles.subItemIcon} />}
      {label && <span>{label}</span>}
      {external && (
        <IconLinkExternal
          size="s"
          className={classNames(styles.icon, styles.verticalAlignSmallOrMediumIcon)}
          aria-hidden
        />
      )}
    </Item>
  );
};

FooterNavigationLink.defaultProps = {
  as: 'a',
};
