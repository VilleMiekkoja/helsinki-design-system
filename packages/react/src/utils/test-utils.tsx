import React, { PropsWithChildren } from 'react';

import { Navigation } from '../components/navigation';
import { Footer } from '../components/footer';
import { Header } from '../components/header';

export const NavigationWrapper = ({ children }: PropsWithChildren<Record<string, unknown>>) => (
  <Navigation menuToggleAriaLabel="menu" skipTo="#content" skipToContentLabel="Skip to content" title="Foo">
    {children}
  </Navigation>
);

export const FooterWrapper = ({ children }: PropsWithChildren<Record<string, unknown>>) => (
  <Footer title="Bar">{children}</Footer>
);

export const FooterNavigationWrapper = ({ children }: PropsWithChildren<Record<string, unknown>>) => (
  <Footer title="Bar">
    <Footer.Navigation>{children}</Footer.Navigation>
  </Footer>
);

export const FooterNavigationGroupsWrapper = ({ children }: PropsWithChildren<Record<string, unknown>>) => (
  <Footer title="Bar">
    <Footer.Navigation>
      <Footer.NavigationGroup>{children}</Footer.NavigationGroup>
    </Footer.Navigation>
  </Footer>
);

export const FooterUtilitiesWrapper = ({ children }: PropsWithChildren<Record<string, unknown>>) => (
  <Footer title="Bar">
    <Footer.Utilities>{children}</Footer.Utilities>
  </Footer>
);

export const HeaderNavigationMenuWrapper = ({ children }: PropsWithChildren<Record<string, unknown>>) => (
  <Header>
    <Header.NavigationMenu>{children}</Header.NavigationMenu>
  </Header>
);
