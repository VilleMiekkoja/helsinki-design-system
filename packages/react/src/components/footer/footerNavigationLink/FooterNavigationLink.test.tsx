import React from 'react';
import { render } from '@testing-library/react';
import { axe } from 'jest-axe';

import { FooterNavigationLink } from './FooterNavigationLink';
import { FooterWrapper } from '../../../utils/test-utils';
import { FooterVariant } from '../Footer.interface';

describe('<Footer.NavigationLink /> spec', () => {
  it('renders the component', () => {
    const { asFragment } = render(<FooterNavigationLink label="Link" />, { wrapper: FooterWrapper });
    expect(asFragment()).toMatchSnapshot();
  });
  it('should not have basic accessibility issues', async () => {
    const { container } = render(<FooterNavigationLink label="Link" />, { wrapper: FooterWrapper });
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
  it('renders the utility variant', () => {
    const { asFragment } = render(<FooterNavigationLink label="Link" variant={FooterVariant.Utility} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders the navigation variant', () => {
    const { asFragment } = render(<FooterNavigationLink label="Link" variant={FooterVariant.Navigation} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders the base variant', () => {
    const { asFragment } = render(<FooterNavigationLink label="Link" variant={FooterVariant.Base} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
