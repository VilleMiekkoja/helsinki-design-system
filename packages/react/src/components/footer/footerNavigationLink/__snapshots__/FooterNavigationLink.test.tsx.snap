// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Footer.NavigationLink /> spec renders the base variant 1`] = `
<DocumentFragment>
  <a
    class="item base"
  >
    <span>
      Link
    </span>
  </a>
</DocumentFragment>
`;

exports[`<Footer.NavigationLink /> spec renders the component 1`] = `
<DocumentFragment>
  <footer
    class="footer koros-basic theme-light"
  >
    <div
      class="koros basic koros basic"
    >
      <svg
        aria-hidden="true"
        height="85"
        width="100%"
        xmlns="http://www.w3.org/2000/svg"
      >
        <defs>
          <pattern
            height="85"
            id="koros_basic-1"
            patternUnits="userSpaceOnUse"
            width="106"
            x="0"
            y="0"
          >
            <path
              d="M0,800h20V0c-4.9,0-5,2.6-9.9,2.6S5,0,0,0V800z"
              transform="scale(5.3)"
            />
          </pattern>
        </defs>
        <rect
          fill="url(#koros_basic-1)"
          height="85"
          width="100%"
        />
      </svg>
    </div>
    <div
      class="footerContent"
    >
      <section
        class="navigationContainer centeredNavigation"
      >
        <div
          class="titleWrapper"
        >
          <h2
            class="title"
          >
            Bar
          </h2>
        </div>
      </section>
      <a
        class="item"
      >
        <span>
          Link
        </span>
      </a>
    </div>
  </footer>
</DocumentFragment>
`;

exports[`<Footer.NavigationLink /> spec renders the navigation variant 1`] = `
<DocumentFragment>
  <a
    class="item navigation"
  >
    <span>
      Link
    </span>
  </a>
</DocumentFragment>
`;

exports[`<Footer.NavigationLink /> spec renders the utility variant 1`] = `
<DocumentFragment>
  <a
    class="item utility"
  >
    <span>
      Link
    </span>
  </a>
</DocumentFragment>
`;
